File: sample-legacy/src/main/java/com/example/legacy/AsyncDataLoader.java
package com.example.legacy;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class AsyncDataLoader {

    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    public interface DataCallback<T> {
        void onSuccess(T result);
        void onError(Exception e);
    }

    public void loadUserData(String userId, DataCallback<User> callback) {
        executor.submit(() -> {
            if (userId == null || userId.trim().isEmpty()) {
                if (callback != null) {
                    callback.onError(new IllegalArgumentException("User ID cannot be null or empty"));
                }
                return;
            }

            try {
                User user = new User(userId, "User_" + userId, userId + "@example.com", 30);
                if (callback != null) {
                    callback.onSuccess(user);
                }
            } catch (Exception e) {
                if (callback != null) {
                    callback.onError(e);
                }
            }
        });
    }

    public void shutdown() {
        executor.shutdown();
    }
}

File: migrated-src/com/example/legacy/AsyncDataLoader.kt
package com.example.legacy

import kotlinx.coroutines.*
import kotlin.coroutines.CoroutineContext

object AsyncDataLoader {
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)

    suspend fun loadUserData(userId: String): Result<com.example.legacy.User> {
        if (userId.isNullOrBlank()) {
            return Result.failure(IllegalArgumentException("User ID cannot be null or empty"))
        }

        try {
            val user = com.example.legacy.User(userId, "User_$userId", "$userId@example.com", 30)
            return Result.success(user)
        } catch (e: Exception) {
            return Result.failure(e)
        }
    }

    fun shutdown() {
        scope.cancel()
    }
}

data class User(
    val id: String,
    val username: String,
    val email: String,
    val age: Int
)